### ISSUE TITLE
Allow modules config to override lack of main file

### ISSUE THREAD

**Issue by chrylis** [dbashford/mimosa-require#44]
I have [an application](https://github.com/artsquare/caniprintit) that's using a number of AMD modules for code-organization and testing purposes. At production build time, I'd like to roll all of these up into a single JavaScript file, where the main entry point is `cipi/main`. (I'm still wanting to include jQuery and Bootstrap from the CDN, so we're using real RequireJS instead of Almond.)

I can't figure out how to communicate to Mimosa what the main module is; I always get `Info - No main modules found.  Not running optimization.` I've tried adding an entry under `require.optimize.modules` in the vein of MimosaDynamicRequire, and that doesn't seem to affect anything, and I didn't connect anything in the docs with the issue. Do you have any suggestions?

**Comment by dbashford** [comment]
Having trouble helping here because I can't figure out how to bring the app up.  Things are a bit all over the place.

Regardless, you don't have a main file anywhere.  All you have defined are a bunch of modules.  Your `main` is embedded in your `.html` file which makes it impossible for Mimosa to discover it.

Here's a more idiomatic way to wire up your app that also help Mimosa help you out.

1) Rename `common.js` to `main.js`, though really the renaming is optional, it'll just make the names make more sense.  Everything will work fine if you don't rename the file.

2) At the bottom of `main.js`, after the `require.config`, put this:  `require(['cipi/main']);`

3) change your `index.html` so that it only brings in require and points at a `data-main`.  After this, remove your other script tags like common and the `require`.  Should just have this one script brought in.
```
<script src="... require.min.js" data-main="js/main"></script>
```

Like I said, can't test this myself.  May have some URLs to monkey with.  IF you do this, then the `require` you add to `main` lets Mimosa know that that file is a main and when you run `mimosa build -om` Mimosa will bundle that for you without you needing to configure anything.

The `modules` config should work, but because Mimosa doesn't discover any mains at all, it doesn't run the `modules` config.  I should probably do something about that, but patching that means you having to write add config and extra `script` tags that are not necessary.

**Comment by dbashford** [comment]
I've updated the title of this to indicate what needs to change.

**Comment by chrylis** [comment]
Sorry, it's a Spring Boot application, and the `resources` submodule is a self-contained Mimosa build (the main build runs `npm install; mimosa build -o` from that directory). A top-level build needs Java 7+, Maven, and Node, and `mvn package` from the root produces a packaged executable.

`common.js` is just the RequireJS configuration and not the actual entry point. I'm using this as a test bed for a much more complicated multiple-page site, and so adding the `require` to that file is a workaround that wouldn't work on a larger scale. Using `data-main` will work, although I don't think that's reliable enough to identify all mains, as we might be supplying that value in the server-side template at runtime.

It seems that I'm unclear on what exactly makes a main module a main module. I had thought that it was simply a top-level entry point unreferenced by other modules (such as how the LESS dependency parser works), but it appears that there's some other criterion for identifying it. I am calling `require(['cipi/main'])` from `index.html`, but that doesn't appear sufficient.

**Comment by dbashford** [comment]
We build multi-single page apps out of the same project directory.  So we have multiple mains.  We do it like this: https://github.com/requirejs/example-multipage/tree/master/www/js

```
require(['./common'], function (common) {
    require(['app/main1']);
});
```

Do it here too: https://github.com/dbashford/MimosaEmberHTMLBarsSkeleton/blob/master/assets/javascripts/blogger-main.js

Having the `main` bring in the common before getting started allows you to have a `common`, but also preserve the nature of a `data-main`.

**Comment by chrylis** [comment]
Okay, so it's the presence of the `require()` that marks a main? We were inlining that in the HTML to avoid an extra JS load, but am I correct that that will all get bundled up?

**Comment by dbashford** [comment]
Yep, all gets bundled up.

You should, just for kicks, run a `mimosa new` and play with the little sample project.  Pick one with hogan views as they are the easiest to read and understand.  You can see how that app is wired together and how, when in optimized mode, the script tag changes.

https://github.com/dbashford/newmimosa/blob/master/skeleton/view/hogan/views/index.hjs

When you bundle everything together in one file you also do not need the `require` config you have: https://github.com/artsquare/caniprintit/blob/master/caniprintit-resources/mimosa-config.coffee#L31-L34

mimosa-require will include Almond.  require.js is big.  It has all the things necessary to do async script loading, which, once everything is in one file, you no longer need.  Almond contains all the is necessary to wire everything together when its all concatenated.

Usually you only need to include require.js in your bundled output if you intend to build several modules and need to preserve async loading.  That's usually only something you'd do for large apps.

**Comment by chrylis** [comment]
Almond doesn't support CDN loading, and the Big Stuff we're using (jQuery, Bootstrap) has shared canonical CDN support.

**Comment by chrylis** [comment]
Is there a way to have Mimosa twiddle the script tag (if necessary) for a plain HTML file? As far as Mimosa is concerned, our HTML files are static HTML.

**Comment by dbashford** [comment]
There's nothing for that no.  But doing it would be fairly easy using adhoc modules if you wanted.  On our production apps we have several adhoc modules to perform some very specific tasks that don't make much sense as published modules.  If you go that route you may find yourself with something worthy of module publishing.

There are ways to compile dynamic ones, but the assumption there is that you are using something during dev that can compile them.  [server-template-compile](https://github.com/dbashford/mimosa-server-template-compile) just does its thing during build.

**Comment by chrylis** [comment]
Are using RequireJS and `modules` mutually exclusive? This configuration block:

    require:
        optimize:
            overrides:
                name: "vendor/requirejs/require"
            modules: [
                { name: "main", exclude: ["printanalyzer/dataTable"] }
            ]

Gives me this error:

    [Error: Error: "name" and "modules" options are incompatible. Either use "name" if doing a single file optimization, or "modules" if you want to target more than one file for optimization.]

**Comment by dbashford** [comment]
I think the idea is that if you are using modules, the `name` is pointless.  Each module needs a name.  That's an error direct from r.js.

If you do want to see exactly the config that mimosa passes to r.js you can run with the `-D` flag.  There is a lot of spam, but towards the end you'll see the r.js params.

**Comment by chrylis** [comment]
Okay. I was attempting to copy MimosaDynamicRequire there.

**Comment by chrylis** [comment]
Aha. I've found something that looks like a problem. I had set `require.optimize.overrides.name` to `vendor/requirejs/require` to use RequireJS instead of Almond, but then that `name` gets passed into the r.js config object. RequireJS then thinks that I'm trying to name the module that it's processing `vendor/requirejs/require`. Removing that override gets me past that bump, but now it's choking on a CDN path for jQuery.

**Comment by dbashford** [comment]
If you are using `modules`, I think the idea is you are not bundling require.js with your application because you are using it to async load modules in?  https://github.com/dbashford/MimosaDynamicRequire/blob/master/views/layout.jade#L11

At this point though you are successfully (I think) getting config through to r.js and if you run into errors it'll be on the r.js side.  It can be fairly picky about how it gets configured which is why I try to do as much as possible via mimosa-require.

**Comment by chrylis** [comment]
`require.optimize.overrides.name` is overloaded: It controls both Mimosa's selection of processor (Almond/RequireJS) *and* is passed to r.js as a configuration option telling r.js to use that name for the constructed module. However, an r.js `name` configuration key is incompatible with a modularized build. It looks like there needs to be a way to tell Mimosa not to use Almond that stays out-of-band.

**Comment by chrylis** [comment]
I'm not certain what you mean by "not bundling", but I'm not wanting to inline it into a chunked JavaScript file. I am async loading a number of distinct modules depending on what the particular pages need.

**Comment by dbashford** [comment]
> require.optimize.overrides.name is overloaded: It controls both Mimosa's selection of processor (Almond/RequireJS) 

Nah, its not overloaded at all.  It's just a pass-through that is set to `almond` by default.  If you don't want almond, you need to set it to something else.  If you provide `modules` it isn't set to anything.

This is all documented: https://github.com/dbashford/mimosa-require#inferred-rjs-optimizer-settings

**Comment by chrylis** [comment]
Hmm. `null` works (just figured that out before I saw that), but still having problems with the CDN. I think this is all in r.js now, so digging there.

**Comment by chrylis** [comment]
Got it. The `common.js` points to the right place, and then CDN entries have to be overridden to be `empty:` in `mimosa-config`. I'll look at putting together a simple example project to demonstrate. In the meantime, this looks like a misfeature in the r.js optimizer.

**Comment by dbashford** [comment]
There is an issue here where the lack of a mimosa-detected main stops module runs.
